### ISSUE TITLE
Can donate/open source a nice THashTree implementation for 2000 !!! pascal coins ! =D

### ISSUE THREAD

**Issue by SkybuckFlying** [PascalCoinDev/PascalCoin#109]
Hello PascalCoin Community reading this !

I was once planning on adding some merkle hash tree to some communication program and therefore I wrote a highly advanced THashTree implementation. I actually also have some other unique classes/designs which might become available later.

But for now let's talk THashTrees ! =D

This PascalCoin project will probably need a THashTree implementation. This is very difficult and error prone code to write. Try googling THashTree and you will only find 2 pages on google ! =D (nothing usefull in those two pages ?!)

THashTree implementation will be necessary for the checkpoint torrenting feature. (Read about it in PascalCoin Paper v2 !)

My work on this feature/idea THashTree/MerkleTrees will go along way at advancing PascalCoin towards having such a feature. I even explored the possibility of "repairing chunks" in case some of them are corrupted, however that would require a tricky/complex protocol. This "repairing" idea could be dropped to keep complexity down for any protocol design. Or Molina could try it, he seems good with protocols ;) But I would advise against adding to many powerfull features in this regard.

Concerning the verification part, that should be easy to implement thanks to this class of mine. It's very suited for that and was exactly design for this.

So before the community/investors/people with lots of pascalcoin say "no to this" or "helllllll noooooo".

Think long and hard about this, because this oppertunity may not come again ! ;) =D

2000 pascalcoins in light of 80.000.000 pascal coins feels like a ripp-off lol ! But at least it allows me and my code to life on forever maybe.... or at least find some nice usage in some nice programs.

Also I would like to see PascalCoin download blocks/chunks faster, thus such a feature is well-comed by me too.

Now that PascalCoin seems to have some "dev" funds... How about shifting some of those funds to me as a long time Delphi developer ?. 

You get.... in this case... immediately... what you pay for.

Let me be clear about this:

5 Test Programs.
1 THashTree unit/class implementation (very advanced can do lot's of things, pruning, recursing, adding, marking and ofcourse calculating hashes, double hashes, etc, all kinds of stuff, I promise you will probably not be disappointed).

What I cannot garantuee:

1. Cannot 100% garantuee suitedness for PascalCoin, I think it is 100% suited, just not entirely sure, it's been a long time since I wrote it. Wrote it in 2008 ! 10 years ago. Some new methods may have to be added, this depends on protocol design, and how it actually wants to download chunks. At least the class will safe Molina a lot of work.

2. Cannot 100% garantuee bug free operations. It's tested extensively/intensely, however such a class remains tricky. I cannot clearly remember any issues, though I do know myself and I do know I put a lot of effort in this class and debugging it, so quality should be pretty high as usual, though this kind of class/operations is tricky. The most important/severe bugs stomped out ofcourse, though 1 or 2 minor bugs may still be present 

3. Cannot garantuee I will perform further work on this class. I have not coded in a long time, my motivation and skills/energy levels are lower than in the past. At the time when I wrote these codes I was at the top of my game though so to speak =D (This would also be my first truely "open source" contributions).

4. My nickname at the very least must remain mentioned in the class, as the original author. Plus date and time of creation/written/development of this class (Hate it when dates are missing =D)
5. My realname must also be mentioned so that in case of my passing I have some nice legacy =D

6. Molina must approve, cause he will probably be the one using the class to implement a torrenting protocol or chunk verification/hash checking protocol something like that.  (I could try and help a little bit with advise and maybe little bits of code, cannot garantuee this though).

7. In case herman or some other programmer is to work on this protocol then this programmer must approve.

At current pascalcoin exchange rate the "fee" payed to me would equal 2000x0.50 euros = 1000 euros.

A bargain for this class I think. Apperently very little programmers around the world are capable of writing this kind of class.

The fee is low for some reasons:

1. I like this project, I want it to advance/succeed.
2. Perhaps some day pascal coins might be worth more than just 50 cents and thus I could get rich that way =D
3. I like to see this class being used in some project.
4. I also wrote about this approach of merkle tree hashing protocol on bitcointalk forum (in 2012), thought about it there some more, after I already wrote this class in 2008. Some on bitcoin talk forum there is some more "documentation/idea" support of mine how to actually implement such a protocol and how to advance this class further if necessary to realize it fully.
5. Having a sense of having made a real (significant/substantial) code contribution to this project.
6. Further code development of the class may be value for one of my other projects some day, gaining more experience with this hashing/protocol idea.

I believe this feature is more difficult/daunting then most would think but I believe that Molina with a little bit of help from me in the form of this class plus some further slight support and advise has a high chance of succeeding implementing this feature.

It could just be the thing that sets PascalCoin apart from other coins. I am no coin expert. But as far as I know no other coins have this feature ? Though perhaps other coins DO have this feature ? Perhaps light clients ? Though perhaps they work differently. Perhaps I or interested readers will inspect other coin systems to see if they already have such a feature.  Copy & Pasting that code and porting it to pascal may then be an alternative option to developing your own.

If anybody knows of another coin system that already has this feature in place, then DO mention it in this issue/thread.

If the response is NAY, then I win too. Cause I will be very curious to see the following:

1. Will this feature ever exist without my help ?!?!?!?
2. And if so.... what will it look like ?

I would be very curious to see alternative implementations. I just hope they don't get too crappy ! =D

Now it's time to be honest with you as well ! I am surprised about Molina's skills ! Very good skills !

This THashTree design of mine was designed to by a "dynamic data structure". This means the class was designed in such a way that it can "reduce memory" by collapsing/pruning the tree.

This particular feature is not required for an initial implementation, but may still be valuable in case multiple trees have to be evaluated/checked etc. However.... just adding these nodes can already get a bit tricky and such, traversing them, etc.

I am not sure if Molina would be capable of recreating such a class. It requires some particular skills:

1. Recursive experience. I am sure he has some of that... all programmers should.
2. Dynamic data structure experience. Here is start to doubt a little bit, but there is some of this in his code.
3. Linking structures together.
4. Great debugging skills.
5. Great test program writting skills =D bit of fun there.

For now I would guess Molina, maybe even Herman is capable of re-creating this class, but also have some doubts. Perhaps some kind of "lazy/array" based implementation may work sufficiently, though if I had to re-create one right now I would scream in horror !

Also I have not thought it completely through, now that I do a little bit there is a further problem with this idea, which basically rules out the "lazy" approach.

The safebox can/will and is already growing quite large. This means many hashes. This could mean that a "DYNAMIC DATA STRUCTURE which can COLLAPSE/BE PRUNED" as hashes that come in allow it, is ABSOLUTELY required, to avoid OUT OF MEMORY situations.

So what I am trying to tell you is.... re-inventing/re-writing this class is going to be a bit PAINFULL and TIME wasting for you !

It's better to "throw away" 2000 pascal coins from your DEV fund and get STELLAR code straight from the bat/start and safe yourself PAIN/BUGS and maybe some embarrasement =D

This is also one of the reasons why I write this.... I am very experienced with writing Delphi code and debugging and I can clearly remember there were some painfull bugs in there... that required really good testing and debugging. I want you to not experience this """PAIN""" =D

I want you to keep FLYING... and write new/good usefull productive code for Pascal Coin ! ;) =D

Stuff/code which I may not be able to write...

My hope is you do suck at this dynamic stuff... that means we can have a good/nice collaboration. Even if you don't suck at it, at least you can safe quite some time.

My current estimation, assuming you do have experience with trees/recursion and such  6 days of programming you could safe.

Yes my estimation is it will take you at least 6 days of pretty intense thinking/programming/debugging to get to the point where my class is now =D

And then you still have to start on the protocol =D

Think about it.... 

Bye for now,
  Skybuck.

**Comment by SkybuckFlying** [comment]
Some other advantages:

Since the code was written in 2008, it should work fine/perfectly in Delphi 2010 ! ;)

Googling HashTree without a capital T in front of it, does produce much more results !

Perhaps there are some nice C classes/libraries or other pieces of code which may come in handy.

I will be honest with you, if I find something which seems way better or more suited then my class I will let you know, otherwise I would feel a bit guilty and responsible for contributing sucking code vs even more stellar/super code =D

Did see some java code, but would rate my code as better/more suited more high performant, based on a couple of seconds of looking at it ! ;) :)

*Update*

GitHub might be your friend (or could be your horror), there could be some c/c++ code there that may be worth something:

https://github.com/search?q=hash+tree&type=Repositories

But I'll be damned if I am gonna try and figure out that C/C++ code ! LOL.

To sweeten the deal a little, know that my Delphi code has almost every line commented (besides from obvious a:=b assignments and such) ! which you will hopefully/probably like much better than some undocumented shady c/c++ code ! =D

Every line of Delphi code is commented so you can follow exactly what is going on inside the code !

This is pretty essential to have some level of reliability and "easier" reasoning about correctness of code ! ;)

This kind of code has to be as flawless as possible, nothing can be left to chance... not even some shady C compiler differences ! ;) (Feel the FUD in this one ? =DDDD) or god forbid (Porting/Language differences between C/C++ tricks and Delphi... feel the SHL fud and such in this one ? =D)

**Comment by SkybuckFlying** [comment]
Finally, let's have a look at the BIG MASTER, BITCOIN ITSELF !

https://github.com/bitcoin/bitcoin/blob/master/src/consensus/merkle.cpp

It starts off with a warning about a flaw:

"The reason is that if the number of hashes in the list at a given time
       is odd, the last one is duplicated before computing the next level (which
       is unusual in Merkle trees). This results in certain sequences of
       transactions leading to the same merkle root. "

In the past I read about this flaw, I think it's a general problem with the merkle tree algorithm idea.

I don't quite understand it, what causes it, and especially how to exploit it. Therefore:

8. I cannot garantuee "exploitation" nor take any responsibility for it. This will have to be looked at by others, and maybe me again, maybe now that I am older and hopefully wiser I can get some more sense of this.

(Anyway this does point out how tricky this all can be... if even the great all allmighty bitcoin trembles in fear ! ;) hihihihhi nice story eh ? ;))

But the real point is, where the hell is it's implementation ?

Well as far as I can tell it uses all kinds of C++ WHICHCRAFT ! LOL... Like "vectors"  and (sexually transmittable diseases STDs lol and then add some)

Though besides the phun at C/C++ it's implementation seems very short and perhaps very efficient (?) at least code wise...

Also it seems to use a stack based approach "push_back" "back"  ? Not sure what that is... some hash implementation, probably related to vector class... could be some array function or so...

I would probably consider this bitcoin class "cheating". It doesn't really implement any true tree like structure ?! Therefore it's probably limited to what it can do.

Though it also seems to use functionality of CBlock. 

This is kinda of the annoying thing about Bitcoin... knowing C/C++ is not enough... must also known these classes which probably come from the even more annoying BOOST library... which requires the latest and greatest C/C++ compiler which can  only be GCC cause all others are only yearly updated... and GCC gets an update whenever ;) Thus no commercial compiler can actually ever compile the latest codes.... and thus open source GCC compiler usage assured ! ;) HAHA !

The include statements also kinda annoying, but I guess this is the same as in uses clausule of Delphi... don't known where hash class resides. Ok found it... it's not in boost... but actually in main folder of bitcoin:

https://github.com/bitcoin/bitcoin/blob/master/src/hash.cpp

Useless does not explain the push_back function... 

Well I give up for now trying to make some sense of this code... with a project file might be a bit better to figure this stuff out... but does it actually have a usuable project file for windows/vc ?  Probably not it does have a makefile:

https://github.com/bitcoin/bitcoin/blob/master/src/Makefile.am

Trying to "decode" this to see where this CBlock file might be ;) actually it's not a file, just some class name or so... hmmm... 

No idea where that could be coming from except maybe the top header...  Surely the great Skybuck can read H files ? hihi.

Found the it:

https://github.com/bitcoin/bitcoin/blob/master/src/primitives/block.h

Fracking useless...

I could go on and inspect Boost... but this is getting very boring ? See my point why I don't like bitcoin code much ?

These freakos... use these crap C++ features and boost crap for the silliest things ? Push back, back really ? Do I really need a HUGE boost library for pesky tiny little features like that ?

Neeeeehhhhh... Got Delphi for that lol HYPOCRIT ME.... but seriously I can write code like that myself don't need booooost for it... 

Will give it one last chance... going to find vector in boost somewhere, first checking:

https://github.com/bitcoin/bitcoin/blob/master/src/uint256.h

Might be in here. Nope... begin and end is in there though.

https://www.boost.org/doc/libs/1_58_0/doc/html/boost/container/vector.html#idp49159008-bb

Yeah push seems to grow an array.

Well anyway... I am going to end my "excursion" into bitcoin crappyness write here ! :P**** love bad mouthing it... but no seriously bitcoin is written quite nicely... but in this particular case I don't know.

BUT NOW COMES SOMETHING IMPORTANT.

Please keep in mind that this particular bitcoin implementation was PROBABLY written for a SINGLE TRANSACTION BLOCK.

This is something totally DIFFERENT than trying to MERKLE TREE the entire SAFEBOX if I understand PascalCoin CHECKPOINTS CORRECTLY.

Therefore IF you were to use this Bitcoin implementation there IS A CHANCE that this implementation MAY RUN OUT OF MEMORY ! =D

FUN FUN FUN FUN FUN ! =D

Maybe not...

You may go ahead and waste your time on trying it out... Maybe you get lucky and it won't run out of memory !

I would be filled with JOY ! LOL =D Not really but a little bit ! ;)

But the smile on your face will probably disappear quite quickly because suddenly you realise... it's USELESS for any advanced protocol stuff !

If you want to do advanced dynamic stuff... or even wild repair dreams... you gonna need something better.

You gonna need Skybuck tech and perhaps some assistance ! ;) =D

Finally you want to beat Bitcoin or NOT ?!

Or do you want to be some half-baked badly working clone without any ground breaking features !? ;) =D

PUT BALLS TO THE WALLS ! LOL And go with the Skybuck TECH ;) =D

It's a bit of a risk... the tech is great... it will make your hart pound a little bit faster ?!

Can you handle the pounding ?! Can you handle the tech ?! Will it forfill your wildest dreams ?!

Can you handle it if it actually does ?!

Programming together can sometimes be difficult with Skybuck cause of dreams coming true... some dreams could turn into AI nightmares very maybe in future... but furtunately that future is still somewhat away....  some dreams are almost to good to be true.... it can be daunting to handle the power !

It's like a light pounding magic sword for real ! Like a light saber... suddenly you have one... what ya gonna do with all that power ?! =D

**Comment by SkybuckFlying** [comment]
I did come across THashTree<T> or something like that and know I found a bug report about this.

It seems to be some generic support build into Delphi.

I am sure Delphi XE probably has this but Delphi 2010 don't know... investigating  ! ;) =D Stay tuned/tuna-ed ! =D (Already checked helpfile but couldn't find anything... but that doesn't mean it's not in the sources somewhere ! ;) fingers crossed ! ;))

Oh.... it's starting to seem this was just a "test case", "custom developed piece of source code" by somebody... and then complaining about generics bugs in the compiler:

For what it's worth it here is the link to that:

https://community.embarcadero.com/images/39973/QC%20Bugs%20Fixed%20in%20Delphi%202010.pdf

(oh cool report about Delphi 2010 and generic fixes ! ;) Yeah... noticed these bugs at the time... they were bluntly obvious... but that's how it goes apperently with this kind of advanced language stuff ;) hmmm well not that advanced or is it hmm...)  

So I am guessing for now... no cookie ! No free THashTree ride ! ;) :)

I will keep searching kinda curious what Delphi XE has to offer in terms of collections and stuff like that. Good turkey filling stuff ! ;)

I also checked this nice Delphi library... looks nice... but very "generic" advanced... not sure if that is a good thing though with all these new potential generics bugs and could be hard to use or perhaps not... not sure if it's Delphi 2010 compatible...

https://github.com/pavkam/DeHL/tree/master/Library/src/Collections

It has a lot... it even has some kind of THashSet... but again no THashTree as far as I can tell.. no cooooookie ! ;) =D Maybe one of the other classes could be used as if it was one TBag maybe ?  guess not though ;)

Well this is a bit disappointing... again no free THashTree coookie ! ;) =D

http://docwiki.embarcadero.com/Libraries/Tokyo/en/System.Generics.Collections

Classes
TArray	Class that contains static methods for searching and sorting a generic array.
TDictionary	Collection of key-value pairs.
TEnumerable	TEnumerable is an abstract class inherited by all enumerable containers.
TEnumerator	TEnumerator is the abstract class implemented by all enumerator classes.
TList	Ordered list.
TObjectDictionary	Collection of key and/or value objects.
TObjectList	Ordered list of objects.
TObjectQueue	Queue of objects.
TObjectStack	Last in, first out stack of objects.
TQueue	Queue implemented over array, using wrapping.
TStack	Last in, first out stack.
TThreadedQueue	Represents a generic threaded queue.
TThreadList	Represents a generic thread-safe list.
Routines
InCircularRange	InCircularRange is an internal function used by the Remove method.
Types
PObject	PObject is a pointer type to a TObject class.
System_Generics_Collections__02	
TCollectionNotification	Type of change to collection for OnNotify event.
TDictionaryOwnerships	Set of ownerships for TObjectDictionary.
TListHelper	Internal helper with non-generic methods to handle TList instances.
TPair	TPair is a record holding a key-value pair.
TQueueHelper	Internal helper with non-generic methods to handle TQueue instances.
TStackHelper	Internal helper with non-generic methods to handle TStack instances.
Package | rtl250.bpl

Well at least there is a stack... not sure if it could be used to mimic bitcoin/boost's vector... which might be a mute point anyway... ofcourse there is always SetLength and such... hmmm...
 
There is TList<T> which is probably more like boost's vector. It can grow and stuff ;) then again the methods can't really do it...

All kinda limited... well that again ends my excursion into it.

Long story short: Nothing beats hand written code =D hihihihihhihi. Old school style.

Who needs these classes and generics anyway ? People that can't program ? Or are in a really bad rush and trying to cut corners the bad way ?! Looks/seems painfull... ouch ! ;) :)
